### PR TITLE
8335059: Consider renaming ClassLoaderData::keep_alive

### PR DESCRIPTION
**Pull request by coleenp**
How does this rename look?  Instead of ClassLoaderData::keep_alive() and a _keep_alive refcount, it's been renamed to _strongly_reachable and is_strongly_reachable().
Tested with tier1 on Oracle supported platforms.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8335059](https://bugs.openjdk.org/browse/JDK-8335059): Consider renaming ClassLoaderData::keep_alive (**Enhancement** - P4)

### Reviewers
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**)
 * [Stefan Karlsson](https://openjdk.org/census#stefank) (@stefank - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/20408/head:pull/20408` \
`$ git checkout pull/20408`

Update a local copy of the PR: \
`$ git checkout pull/20408` \
`$ git pull https://git.openjdk.org/jdk.git pull/20408/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 20408`

View PR using the GUI difftool: \
`$ git pr show -t 20408`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/20408.diff">https://git.openjdk.org/jdk/pull/20408.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/20408#issuecomment-2261174845)

### CODE REVIEW

**Comment by stefank** [comment]
There's a risk that someone incorrectly interprets:
```
- _strongly_reachable 0
```

to mean that the class loader isn't strongly reachable.

In the bug entry I suggested a name `_strong_count` and tried to avoid the word "reachable" because it already have a meaning for the GC. When talking about the "strong" property we both talk about strongly reachable and strong roots. The property for this CLD is that it is a strong root from the GC's perspective. Maybe we can use that instead. What do you think about 
`_strong_root` and `is_strong_root`? Or maybe even `_root` and `is_root`?

File: src/hotspot/share/classfile/classLoaderDataGraph.cpp [hunk 1]
```diff
@@ -310,7 +310,7 @@ void ClassLoaderDataGraph::modules_do_keepalive(void f(ModuleEntry*)) {
   ClassLoaderDataGraphIterator iter;
   while (ClassLoaderData* cld = iter.get_next()) {
     // Keep the holder alive.
-    (void)cld->holder();
+    cld->keep_alive();
```
Comment (dholmes-ora) [inline src/hotspot/share/classfile/classLoaderDataGraph.cpp:313]: Does comment need changing or removing?

File: src/hotspot/share/classfile/classLoaderDataGraph.cpp [hunk 2]
```diff
@@ -335,7 +335,7 @@ void ClassLoaderDataGraph::loaded_classes_do_keepalive(KlassClosure* klass_closu
...
```
Comment (dholmes-ora) [inline src/hotspot/share/classfile/classLoaderDataGraph.cpp:338]: Ditto re comment

**Review by dholmes-ora — COMMENTED** [review]
At the moment the changes do not at all match the PR description, and while the renaming adds a little clarity it doesn't seem  to address any concerns one might have had with "keep alive" as a terminology.

> I want the attribute to tell me that GC can't unload this CLD!

Maybe `disable_unload_count`?

File: src/hotspot/share/classfile/classLoaderData.hpp [hunk 3]
```diff
@@ -205,12 +205,14 @@ class ClassLoaderData : public CHeapObj<mtClass> {
...
   oop holder_no_keepalive() const;
+  // Resolving the holder keeps this CLD alive for the current GC cycle.
   oop holder() const;
+  void keep_alive() const { (void)holder(); }
 
   void classes_do(void f(Klass* const));
 
  private:
-  bool keep_alive() const       { return _keep_alive > 0; }
+  bool keep_alive_ref_count() const { return _keep_alive_ref_count; }
```
Comment (stefank) [inline src/hotspot/share/classfile/classLoaderData.hpp:215]: keep_alive_ref_count() should return an int. The Windows GHA build complains about this.

File: src/hotspot/share/classfile/classLoaderData.hpp [hunk 4]
```diff
@@ -303,8 +305,8 @@ class ClassLoaderData : public CHeapObj<mtClass> {
   }
 
   // Used to refcount a non-strong hidden class's s CLD in order to indicate their aliveness.
-  void inc_keep_alive();
-  void dec_keep_alive();
+  void inc_keep_alive_ref_count();
```
Comment (stefank) [inline src/hotspot/share/classfile/classLoaderData.hpp:308]: This comment is slightly misleading. We have two mechanisms to determine a CLDs aliveness. The first is this ref-counting mechanism, the other is the GC tracing. One can misinterpret this comment to mean that we only use the ref-count to determine the aliveness.

Could we tweak this comment to say something that the ref-count is used to *force* the aliveness of CLDs?

File: src/hotspot/share/classfile/classLoaderDataGraph.cpp (hunk 1 above)
Comment (stefank) [reply src/hotspot/share/classfile/classLoaderDataGraph.cpp:313]: I agree. There's no need to talk about the holder here.

**Review by stefank — COMMENTED** [review]
Me and Coleen talked about and agreed on the keep_alive_ref_count name. I think this looks good. I'll Approve the PR when the Windows compilation error has been fixed. Thanks!

File: src/hotspot/share/classfile/classLoaderData.hpp (hunk 3 above)
Comment (coleenp) [reply src/hotspot/share/classfile/classLoaderData.hpp:215]: I don't know why the linux compiler I used didn't, but fixed.

File: src/hotspot/share/classfile/classLoaderData.hpp (hunk 4 above)
Comment (coleenp) [reply src/hotspot/share/classfile/classLoaderData.hpp:308]: How about:
    // Used to refcount a non-strong hidden class's CLD in order to force its aliveness during
    // loading, when gc tracing may not find this CLD alive through the holder.

**Review by stefank — APPROVED** [review]

**Review by dholmes-ora — APPROVED** [review]
If @stefank is happy with this then I am too. Looks good. Thanks for the updates.

**Comment by coleenp** [comment]
Thanks for the review, David and discussion and review, Stefan.
/integrate

**Pull request closed without merging**